python/examples/math/program.py

```diff
@@ -0,0 +1,166 @@
+from __future__ import annotations
+import asyncio
+import json
+from textwrap import dedent, indent
+from typing import Mapping, TypeVar, Any, Callable, Awaitable, TypedDict, Annotated,  NotRequired, override
+
+from typechat import (
+    Failure,
+    Result,
+    Success,
+    TypeChatModel,
+    TypeChatValidator,
+    TypeChatTranslator,
+    python_type_to_typescript_schema,
+)
+import collections.abc
+
+T = TypeVar("T", covariant=True)
+
+def Doc(s: str) -> str:
+    return s
+
+
+program_schema_text = '''
+// A program consists of a sequence of function calls that are evaluated in order.
+export type Program = {
+    "@steps": FunctionCall[];
+}
+
+// A function call specifies a function name and a list of argument expressions. Arguments may contain
+// nested function calls and result references.
+export type FunctionCall = {
+    // Name of the function
+    "@func": string;
+    // Arguments for the function, if any
+    "@args"?: Expression[];
+};
+
+// An expression is a JSON value, a function call, or a reference to the result of a preceding expression.
+export type Expression = JsonValue | FunctionCall | ResultReference;
+
+// A JSON value is a string, a number, a boolean, null, an object, or an array. Function calls and result
+// references can be nested in objects and arrays.
+export type JsonValue = string | number | boolean | null | { [x: string]: Expression } | Expression[];
+
+// A result reference represents the value of an expression from a preceding step.
+export type ResultReference = {
+    // Index of the previous expression in the "@steps" array
+    "@ref": number;
+};
+'''
+
+
+ResultReference = TypedDict(
+    "ResultReference", {"@ref": Annotated[int, Doc("Index of the previous expression in the 'steps' array")]}
+)
+
+FunctionCall = TypedDict(
+    "FunctionCall",
+    {
+        "@func": Annotated[str, Doc("Name of the function")],
+        "@args": NotRequired[Annotated[list["Expression"], Doc("Arguments for the function, if any")]],
+    },
+)
+
+JsonValue = str | int | float | bool | None | dict[str, "Expression"] | list["Expression"]
+Expression = JsonValue | FunctionCall | ResultReference
+
+JsonProgram = TypedDict("Program", {"@steps": list[FunctionCall]})
+
+
+async def evaluate_json_program(program: Mapping[str, Any], onCall: Callable[[str, list[Any]], Awaitable[Any]]) -> Any:
+    results: list[Any] = []
+
+    async def evaluate_array(array: list[Any]) -> list[Any]:
+        return await asyncio.gather(*[evaluate_call(e) for e in array])
+
+    async def evaluate_object(expr: FunctionCall):
```

Comment:
would be good to define the return types expected for this function like maybe Union[List[str], ... , None]: